pkg/app/grp_curl.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package app
+
+import (
+	"github.com/maistra/maistra-test-tool/pkg/util/oc"
+	"github.com/maistra/maistra-test-tool/pkg/util/test"
+)
+
+type grpcurl struct {
+	ns         string
+	securePort string
+	secureHost string
+}
+
+var _ App = &grpcurl{}
+
+func GrpCurl(ns string, securePort string, secureHost string) App {
+
+	return &grpcurl{
+		ns:         ns,
+		securePort: securePort,
+		secureHost: secureHost,
+	}
+}
+
+func (a *grpcurl) Name() string {
+	return "grpcurl"
+}
+
+func (a *grpcurl) Namespace() string {
+	return a.ns
+}
+
+func (a *grpcurl) Install(t test.TestHelper) {
+	t.T().Helper()
+	oc.ApplyTemplate(t, a.ns, grpcCurlTemplate, a.values())
+}
+
+func (a *grpcurl) Uninstall(t test.TestHelper) {
+	t.T().Helper()
+	oc.DeleteFromTemplate(t, a.ns, grpcCurlTemplate, nil)
+}
+
+func (a *grpcurl) WaitReady(t test.TestHelper) {
+	t.T().Helper()
+	oc.WaitDeploymentRolloutComplete(t, a.ns, "grpcurl")
+}
+
+func (a *grpcurl) values() map[string]interface{} {
+	return map[string]interface{}{
+		"SecurePort": a.securePort,
+		"SecureHost": a.secureHost,
+	}
+}
+
+const grpcCurlTemplate = `
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: grpcurl
+spec:
+  replicas: 1
+  template:
+    metadata:
+      labels:
+        app: grpcurl
+        version: v1
+    spec:
+      containers:
+      - name: grpcurl
+        image: {{ image "grpcurl" }}
+        imagePullPolicy: IfNotPresent
+        command: ["sh", "-c"]
+        args: 
+        - |
+          echo "Empty command for grpc service to be ready"
+          grpcurl -insecure -servername grpc.example.com {{ .SecureHost }}:{{ .SecurePort }} list
```

Comment:
I figured out that `-servername` will be ignored in this case and it's used for validating the TLS certificate, not for setting SNI as I expected.
```
  -servername string
    	Override server name when validating TLS certificate. This flag is
    	ignored if -plaintext or -insecure is used.
    	NOTE: Prefer -authority. This flag may be removed in the future. It is
    	an error to use both -authority and -servername (though this will be
    	permitted if they are both set to the same value, to increase backwards
    	compatibility with earlier releases that allowed both to be set).
  -authority string
    	The authoritative name of the remote server. This value is passed as the
    	value of the ":authority" pseudo-header in the HTTP/2 protocol. When TLS
    	is used, this will also be used as the server name when verifying the
    	server's certificate. It defaults to the address that is provided in the
    	positional arguments, or 'localhost' in the case of a unix domain
    	socket.
```
I think we could try to use `-authority` flag to remove the Route:
```suggestion
          grpcurl -insecure -authority grpc.example.com {{ .SecureHost }}:{{ .SecurePort }} list
```
and try to set the ingress gateway FQDN as `.SecureHost`. I guess its resolved IP address should be used to connect and the authority will be used as SNI in the TLS handshake.

Reply:
Done.